REGRESSION (282704@main): [ macOS wk2 ] media/media-vp8-hiddenframes.html is a flaky image only failure.

jyavenard opened the pull request:
#### 2cdde9cb8a0dad7c36895cb190d93922b6cd5dab
<pre>
REGRESSION (282704@main): [ macOS wk2 ] media/media-vp8-hiddenframes.html is a flaky image only failure.
<a href="https://bugs.webkit.org/show_bug.cgi?id=278833">https://bugs.webkit.org/show_bug.cgi?id=278833</a>
<a href="https://rdar.apple.com/134903291">rdar://134903291</a>

Reviewed by Youenn Fablet.

This commit handles two issues.
1- The test incorrectly failed as we weren&apos;t always ending the test once the last frame was displayed as expected. For now we will simply wait for 1s
2- On Apple platforms other than mac, decoding of videos with alternate&apos;s frame was broken. The WebM demuxer pack all CMSampleBuffer with the same timecode
together so that they can be added to the MSE&apos;s TrackBuffer without issues (as we use the time as key). While the VideoToolbox decoder handled
thos CMSampleBuffer properly, the WebCoreDecompressionSession&apos;s VideoDecoder didn&apos;t. We now iterate over all the inner samples and return the last
deocded frames returned.

* LayoutTests/TestExpectations:
* LayoutTests/media/media-source/media-source-vp8-hiddenframes.html:
* LayoutTests/media/media-vp8-hiddenframes.html:
* LayoutTests/platform/mac-wk2/TestExpectations: Remove failure expectations.
* Source/WebCore/platform/VideoDecoder.h: Add enum to treat no frame as error. This is done to leave the current behaviour of WebCodec which does expect an error. To be further investigated in webkit.org/b/282227
Fly-By, re-order the enum to follow coding style (No first)
* Source/WebCore/platform/graphics/cocoa/WebCoreDecompressionSession.mm:
(WebCore::WebCoreDecompressionSession::decodeSampleInternal):
(WebCore::WebCoreDecompressionSession::initializeVideoDecoder):
* Source/WebCore/platform/libwebrtc/LibWebRTCVPXVideoDecoder.cpp:
(WebCore::LibWebRTCVPXInternalVideoDecoder::decode):
(WebCore::LibWebRTCVPXInternalVideoDecoder::LibWebRTCVPXInternalVideoDecoder):

Canonical link: <a href="https://commits.webkit.org/286329@main">https://commits.webkit.org/286329@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/343738581f2a1fa1e38cb43ff0a5b3ebcb68046e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/75661 "Passed style check") | [⏳ 🛠 ios ](https://ews-build.webkit.org/#/builders/iOS-18-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/28512 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/80147 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/26925 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/77777 "Passed tests") | [⏳ 🛠 ios-sim ](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/2949 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/59337 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/17512 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/78728 "Passed tests") | [⏳ 🧪 ios-wk2 ](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/64987 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/39700 "Passed tests") | 
| | [⏳ 🧪 ios-wk2-wpt ](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/22471 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/25254 "Built successfully") | 
| | [⏳ 🧪 api-ios ](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/22807 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/81620 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/3000 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/1889 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/67570 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/3151 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/64960 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/66877 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16667 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/10826 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/8972 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/2957 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/5779 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/2982 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/3917 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/2989 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on LayoutTests/TestExpectations:
```diff
@@ -5670,7 +5670,6 @@ imported/w3c/web-platform-tests/webstorage/document-domain.html [ Skip ]
 
 webkit.org/b/221973 media/media-webm-no-duration.html [ Skip ]
 webkit.org/b/222493 media/media-source/media-source-webm-vp8-malformed-header.html [ Skip ]
-media/media-source/media-source-vp8-hiddenframes.html [ Skip ] # Requires VP8 decoder
```
Seems to fail in WK1

jyavenard replied:
yes, I need to add the exception there, no vp8 decoder for now

youennf commented on Source/WebCore/platform/VideoDecoder.h:
```diff
@@ -39,14 +39,16 @@ class VideoDecoder {
-    enum class HardwareBuffer { Yes, No };
+    enum class HardwareAcceleration { No, Yes };
+    enum class HardwareBuffer { No, Yes };
+    enum class TreatNoOutputAsError { No, Yes };
```
Let's add `: bool` to all these

youennf commented on Source/WebCore/platform/graphics/cocoa/WebCoreDecompressionSession.mm:
```diff
@@ -379,50 +379,47 @@
-                    producer.reject(m_lastDecodingError);
+                assertIsCurrent(protectedThis->m_decompressionQueue.get());
+                if (protectedThis->m_lastDecodingError != noErr)
+                    producer.reject(protectedThis->m_lastDecodingError);
```
How about `if () { reject; return;`.

jyavenard committed:
REGRESSION (282704@main): [ macOS wk2 ] media/media-vp8-hiddenframes.html is a flaky image only failure.
https://bugs.webkit.org/show_bug.cgi?id=278833
rdar://134903291

Reviewed by Youenn Fablet.

This commit handles two issues.
1- The test incorrectly failed as we weren't always ending the test once the last frame was displayed as expected. For now we will simply wait for 1s
2- On Apple platforms other than mac, decoding of videos with alternate's frame was broken. The WebM demuxer pack all CMSampleBuffer with the same timecode
together so that they can be added to the MSE's TrackBuffer without issues (as we use the time as key). While the VideoToolbox decoder handled
thos CMSampleBuffer properly, the WebCoreDecompressionSession's VideoDecoder didn't. We now iterate over all the inner samples and return the last
deocded frames returned.

* LayoutTests/TestExpectations:
* LayoutTests/media/media-source/media-source-vp8-hiddenframes.html:
* LayoutTests/media/media-vp8-hiddenframes.html:
* LayoutTests/platform/mac-wk2/TestExpectations: Remove failure expectations.
* Source/WebCore/platform/VideoDecoder.h: Add enum to treat no frame as error. This is done to leave the current behaviour of WebCodec which does expect an error. To be further investigated in webkit.org/b/282227
Fly-By, re-order the enum to follow coding style (No first)
* Source/WebCore/platform/graphics/cocoa/WebCoreDecompressionSession.mm:
(WebCore::WebCoreDecompressionSession::decodeSampleInternal):
(WebCore::WebCoreDecompressionSession::initializeVideoDecoder):
* Source/WebCore/platform/libwebrtc/LibWebRTCVPXVideoDecoder.cpp:
(WebCore::LibWebRTCVPXInternalVideoDecoder::decode):
(WebCore::LibWebRTCVPXInternalVideoDecoder::LibWebRTCVPXInternalVideoDecoder):

Canonical link: https://commits.webkit.org/286329@main
